server/app/import.go

```diff
@@ -0,0 +1,172 @@
+package app
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"path"
+	"path/filepath"
+
+	"github.com/krolaw/zipstream"
+
+	"github.com/mattermost/focalboard/server/model"
+	"github.com/mattermost/focalboard/server/services/store"
+	"github.com/mattermost/focalboard/server/utils"
+
+	"github.com/mattermost/mattermost-server/v6/shared/mlog"
+)
+
+const (
+	archiveVersion = 2
+)
+
+// ImportArchive imports an archive containing zero or more boards, plus all
+// associated content, including cards, content blocks, views, and images.
+//
+// Archives are ZIP files containing a `version.json` file and zero or more
+// directories, each containing a `board.jsonl` and zero or more image files.
+func (a *App) ImportArchive(r io.Reader, opt model.ImportArchiveOptions) error {
+	zr := zipstream.NewReader(r)
+
+	boardMap := make(map[string]string) // maps old board ids to new
+
+	for {
+		hdr, err := zr.Next()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				a.logger.Debug("import archive - done", mlog.Int("boards_imported", len(boardMap)))
+				return nil
+			}
+			return err
+		}
+
+		dir, filename := path.Split(hdr.Name)
+		dir = path.Clean(dir)
+
+		switch filename {
+		case "version.json":
+			ver, errVer := parseVersionFile(zr)
+			if errVer != nil {
+				return errVer
+			}
+			if ver != archiveVersion {
+				return model.NewErrUnsupportedArchiveVersion(ver, archiveVersion)
+			}
+		case "board.jsonl":
+			boardID, err := a.ImportBoardJSONL(zr, opt)
+			if err != nil {
+				return fmt.Errorf("cannot import board %s: %w", dir, err)
+			}
+			boardMap[dir] = boardID
+		default:
+			// import file/image;  dir is the old board id
+			boardID, ok := boardMap[dir]
```

Comment:
Can this generate problem if I manually generate this not respecting the expected order on the zip file creation?

I mean, the `boardMap` is populated on the cases where the filename is `board.jsonl` and I guess that is done by finding that file in the zip file in the right order, but If the order is not the right one?

Maybe we should create the boardID before the loop and allow to pass to the `ImportBoardJSONL` an option with the BoardID, that way we would have the new boardID set no matter if the `board.jsonl` has been processed or not.

Reply:
We are using a streaming file format so the order is guaranteed and helps us manage the memory footprint.  That fact it is zip is incidental.  That said, I can do what you suggest to allow different ordering of files in a board directory.  We cannot handle someone changing the order of contents in the `board.jsonl`.

Reply:
We have to decide what is acceptable and what is not. Maybe we can set the limit with things like "order inside boards.jsonl" or define that the zip file needs to add the files in the correct order, but we should be clear upfront with that.

Reply:
On closer look this is tricky without a streaming version of `GenerateBlockIDs`.  That generates a new board id and I won't know what that new board id is if a file appears first.  That is needed to write out the file.